src/sync/partial_sync.cpp

```diff
@@ -56,6 +151,123 @@ void update_schema(Group& group, Property matches_property)
+
+    auto& work_queue = _impl::RealmCoordinator::get_coordinator(config)->partial_sync_work_queue();
+    work_queue.enqueue([object_type=std::move(object_type), query=std::move(query), name=std::move(name),
+                        callback=std::move(callback), config=std::move(config)] {
+        try {
+            with_open_shared_group(config, [&](SharedGroup& sg) {
+                _impl::WriteTransactionNotifyingSync write(config, sg);
+
+                auto matches_property = std::string(object_type) + "_matches";
+
+                auto table = ObjectStore::table_for_object_type(write.get_group(), result_sets_type_name);
+                ResultSetsColumns columns(*table, matches_property);
+
+                // Update schema if needed.
+                if (columns.matches_property == npos) {
+                    auto target_table = ObjectStore::table_for_object_type(write.get_group(), object_type);
+                    columns.matches_property = table->add_column_link(type_LinkList, matches_property, *target_table);
+                } else {
+                    // FIXME: Validate that the column type and link target are correct.
+                }
+
+                if (!validate_existing_subscription(*table, columns, name, query, matches_property)) {
+                    auto row_ndx = sync::create_object(write.get_group(), *table);
+                    table->set_string(columns.name, row_ndx, name);
+                    table->set_string(columns.query, row_ndx, query);
+                    table->set_string(columns.matches_property_name, row_ndx, matches_property);
+                }
+
+                write.commit();
+            });
+        } catch (...) {
+            callback(std::current_exception());
+            return;
+        }
+
+        callback(nullptr);
+    });
+}
+
+void enqueue_unregistration(Object result_set, std::function<void()> callback)
+{
+    auto realm = result_set.realm();
+    auto config = realm->config();
+    auto& work_queue = _impl::RealmCoordinator::get_coordinator(config)->partial_sync_work_queue();
+
+    // Export a reference to the __ResultSets row so we can hand it to the worker thread.
+    // We store it in a shared_ptr as it would otherwise prevent the lambda from being copyable,
+    // which `std::function` requires.
+    auto handover = std::make_shared<_impl::RowHandover>(*realm, result_set.row());
+
+    work_queue.enqueue([handover=std::move(handover), callback=std::move(callback),
+                        config=std::move(config)] () {
+        with_open_shared_group(config, [&](SharedGroup& sg) {
+            // Import handed-over object.
+            sg.begin_read(handover->version);
+            Row row = *sg.import_from_handover(std::move(handover->row));
+            sg.unpin_version(handover->version);
+
+            _impl::WriteTransactionNotifyingSync write(config, sg);
+            if (row.is_attached())
+                row.move_last_over();
```

Comment:
Could do `if (row.is_attached()) { row.move_last_over(); write.commit(); } else { write.cancel(); }` to avoid spurious writes if unregister is called twice.

But meh, maybe not worth caring even that small amount about that.